telemetry/jetbrains/src/main/kotlin/software/aws/toolkits/telemetry/generator/TelemetryGenerator.kt

```diff
@@ -15,142 +16,199 @@ import com.squareup.kotlinpoet.TypeSpec
-        val additionalParameters = metric.metadata?.map { metadata ->
-            val telemetryMetricType = types.find { it.name == metadata.type }
-                ?: throw IllegalStateException("Type ${metadata.type} on ${metric.name} not found in types!")
-            val typeName = if (telemetryMetricType.allowedValues != null) {
-                ClassName(PACKAGE_NAME, telemetryMetricType.name.toTypeFormat())
-            } else {
-                telemetryMetricType.type?.getTypeFromType() ?: com.squareup.kotlinpoet.STRING
-            }.copy(nullable = metadata.required == false)
-
-            val parameterSpec = ParameterSpec.builder(telemetryMetricType.name.toArgumentFormat(), typeName)
-            if (metadata.required == false) {
-                parameterSpec.defaultValue("null")
-            }
-            parameterSpec.build()
-        } ?: listOf()
-        functionBuilder
-            .addParameter(ParameterSpec.builder("project", projectParameter).defaultValue("null").build())
-            .addParameters(additionalParameters)
-            .addParameter(ParameterSpec.builder("value", DOUBLE).defaultValue("1.0").build())
-            .addParameter(ParameterSpec.builder("createTime", Instant::class).defaultValue("Instant.now()").build())
-    }
+    return this
+}
 
-    private fun generateFunctionBody(functionBuilder: FunSpec.Builder, metric: Metric) {
-        val telemetryClient = MemberName("software.aws.toolkits.jetbrains.services.telemetry", "TelemetryService")
-        val metricUnit = MemberName("software.amazon.awssdk.services.toolkittelemetry.model", "Unit")
-        functionBuilder
-            .addStatement("%M.getInstance().record(project) { ", telemetryClient)
-            .addStatement("datum(%S) {", metric.name)
-            .addStatement("createTime(createTime)")
-            .addStatement("unit(%M.${(metric.unit ?: MetricUnit.NONE).name})", metricUnit)
-            .addStatement("value(value)")
-        metric.metadata?.forEach {
-            if (it.required == false) {
-                functionBuilder.beginControlFlow("if(%L != null) {", it.type.toArgumentFormat())
-            }
-            functionBuilder.addStatement("metadata(%S, %L.toString())", it.type.toArgumentFormat(), it.type.toArgumentFormat())
-            if (it.required == false) {
-                functionBuilder.endControlFlow()
-            }
-        }
-        functionBuilder.addStatement("}}")
-        functionBuilder.addKdoc(metric.description)
+private fun buildParameters(metric: Metric, types: List<TelemetryMetricType>): List<ParameterSpec> {
+    val projectParameter = ClassName("com.intellij.openapi.project", "Project").copy(nullable = true)
+    val list = mutableListOf<ParameterSpec>()
+    list.add(ParameterSpec.builder("project", projectParameter).defaultValue("null").build())
+    list.addAll(metric.metadata?.map { metadata -> metadata.metadataToParameter(types) } ?: listOf())
```

Comment:
nitty given how minor it is but more of an FYI
`source?.map { it.toString() }?.toCollection(list)` can prevent tou from making the ?: and the intermediate list

Reply:
Hm the IDE still says that is nullable,  it looks like toCollection doesn't allow null receiver?